### PR TITLE
add low level APIs

### PR DESCRIPTION
**Pull request by GTwhy**

### CODE REVIEW

File: src/macro_utilities.rs [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+            }
+        }
+    };
+}
+
+/// Convert different `BuilderError`s.
+/// Impl `From<XBuilderError>` for `YBuilderError`.
+#[macro_export]
+macro_rules! impl_from_buidler_error_for_another {
+    ($x:path ,$y:path) => {
```
Comment (rogercloud) [inline src/macro_utilities.rs:19]: The same as above

File: src/macro_utilities.rs [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+            }
+        }
+    };
+}
+
+/// Sub builder setter methord helper.
+/// Impl `From<XBuilderError>` for `YBuilderError`.
+#[macro_export]
+macro_rules! impl_sub_builder_setter {
+    ($builder:path ,$sub_builder:path) => {
```
Comment (rogercloud) [inline src/macro_utilities.rs:33]: The same as above

**Review by rogercloud — COMMENTED** [review]
Do you need to add an example to show how to use low level API?

**Comment by GTwhy** [comment]
> Do you need to add an example to show how to use low level API?

Most of them are setter APIs for Builder to change one of qp's attributes, and I think they are easy to understand if users want to change some configurations.
For the complex ones, I added example for them : )

**Review by rogercloud — APPROVED** [review]

**Pull request merged**